### PR TITLE
Jwoo/ios media

### PR DESCRIPTION
**Pull request by nesalang**
Add support for media playback.
Supported media: mp4/mp3/mpeg
Poster is used as placeholder image until user clicks it.
clicking the image plays media on AVPlayerViewController.
playicon image is also supported from hostconfig; otherwise default play icon is drawn in the sublayer.

### CODE REVIEW

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRAVPlayerViewHoldingUIView.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//
+//  ACRAVPlayerViewHoldingUIView
+//  ACRAVPlayerViewHoldingUIView.h
+//
+//  Copyright © 2017 Microsoft. All rights reserved.
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRAVPlayerViewHoldingUIView.h]: change to 2018

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 2]
```diff
@@ -0,0 +1,130 @@
+//
+//  ACRMediaRenderer
+//  ACRMediaRenderer.mm
+//
+//  Copyright © 2018 Microsoft. All rights reserved.
+//
+
+#import <CoreGraphics/CoreGraphics.h>
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: check headers

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        hostConfig:(ACOHostConfig *)acoConfig;
+{
+    std::shared_ptr<BaseCardElement> elem = [acoElem element];
+    std::shared_ptr<Media> mediaElem = std::dynamic_pointer_cast<Media>(elem);
+
+    NSMutableDictionary *imageViewMap = [rootView getImageMap];
+    NSString *key = [NSString stringWithCString:mediaElem->GetPoster().c_str() encoding:[NSString defaultCStringEncoding]];
+    UIImage *img = imageViewMap[key];
+    ACRAVPlayerViewHoldingUIView *view = nil;
+    CGFloat heightToWidthRatio = 0.0f;
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: add more comments

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    
+    NSString *piikey = [NSString stringWithCString:[acoConfig getHostConfig]->media.playButton.c_str() encoding:[NSString defaultCStringEncoding]];
+    UIImage *playIconImage = imageViewMap[piikey];
+    UIImageView *playIconImageView = nil;
+    BOOL drawDefaultPlayIcon = YES;
+    
+    if(playIconImage) {
+        drawDefaultPlayIcon = NO;
+        playIconImageView = [[UIImageView alloc]initWithImage:playIconImage];
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: space

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        playIconImageView.tag = playIconTag;
+        playIconImageView.translatesAutoresizingMaskIntoConstraints = NO;
+    }
+    
+    view.frame = CGRectMake(0, 0, viewGroup.frame.size.width, viewGroup.frame.size.width * heightToWidthRatio);
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: this one looks redundant when default poster is used.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    view.tag = posterTag;
+    ACRContentHoldingUIView *contentholdingview = [[ACRContentHoldingUIView alloc] initWithFrame:view.frame];
+    
+    if(!drawDefaultPlayIcon) {
+        view.hidePlayIcon = YES;
+        [view setNeedsLayout];
+        [view addSubview:playIconImageView];
+
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: remove space

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+- (void)doSelectAction
+{
+    if(!isInline) {
+        if([_view.mediaDelegate respondsToSelector:@selector(didFetchMediaEvent: card:)]) {
+            [_view.mediaDelegate didFetchMediaEvent:_mediaEvent card:[_view card]];
+        } else {
+            // TODO: implemented as a renderer warning -joswo
+            NSLog(@"Warning: inline media play is disabled and host doesn't handles media event");
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: remove s

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+                    UIView *mediaView = self->_mediaViewController.view;
+                    mediaView.frame = frame;
+                    mediaView.translatesAutoresizingMaskIntoConstraints = NO;
+                    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:frame.size.width].active = YES;
+                    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:frame.size.height].active = YES;
+
+                    ACRAVPlayerViewHoldingUIView *poster = [self->_containingview viewWithTag:posterTag];
+                    poster.hidePlayIcon = YES;
+
+                    [poster layoutSubviews];
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: try it with setneedlayoutsubviews

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    
+    ACRAVPlayerViewHoldingUIView *poster = [self->_containingview viewWithTag:posterTag];
+    poster.hidden = YES;
+    
+    [self->_containingview addSubview:mediaView];
+    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self->_containingview attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0].active = YES;
+    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self->_containingview attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:0].active = YES;
+
+    if([_view.mediaDelegate respondsToSelector:@selector(didFetchMediaViewController: card:)]){
+        [parentViewController didMoveToParentViewController:self->_mediaViewController];
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: change condition to check if parentViewController is nil

**Comment by bviglietta** [comment]
Please include a description of the changes in the pull request.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self->_containingview attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:0].active = YES;
+
+                    overlayview.backgroundColor = UIColor.blackColor;
+                    // overlayview sit between AVPLayverViewController view's control view and content view, and here we add the poster.
+                    [overlayview addSubview:poster];
+                    [overlayview bringSubviewToFront:poster];
+
+                    [NSLayoutConstraint activateConstraints:@[[poster.topAnchor constraintEqualToAnchor:overlayview.topAnchor], [poster.bottomAnchor constraintEqualToAnchor:overlayview.bottomAnchor], [poster.leadingAnchor constraintEqualToAnchor:overlayview.leadingAnchor], [poster.trailingAnchor constraintEqualToAnchor:overlayview.trailingAnchor]]];
+
+                    [parentViewController didMoveToParentViewController:self->_mediaViewController];
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: >[parentViewController didMoveToParentViewController:self->_mediaViewController]; [](start = 20, length = 80)

This has the parent and child view controllers reversed.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    
+    ACRAVPlayerViewHoldingUIView *poster = [self->_containingview viewWithTag:posterTag];
+    poster.hidden = YES;
+    
+    [self->_containingview addSubview:mediaView];
+    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self->_containingview attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0].active = YES;
+    [NSLayoutConstraint constraintWithItem:mediaView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self->_containingview attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:0].active = YES;
+
+    if(parentViewController){
+        [parentViewController didMoveToParentViewController:self->_mediaViewController];
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: >[parentViewController didMoveToParentViewController:self->_mediaViewController]; [](start = 8, length = 80)

This has the parent and child view controllers reversed.

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ViewController.m [hunk 12]
```diff
@@ -280,6 +281,12 @@ - (void) didFetchUserResponses:(ACOAdaptiveCard *)card action:(ACOBaseActionElem
 - (void)didFetchSecondaryView:(ACOAdaptiveCard *)card navigationController:(UINavigationController *)navigationController{
     [self presentViewController:navigationController animated:YES completion:nil];
 }
+
+- (UIViewController *)didFetchMediaViewController:(AVPlayerViewController *)controller card:(ACOAdaptiveCard *)card {
+    [self addChildViewController:controller];
+    //[self presentViewController:controller animated:YES completion:nil];
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ViewController.m]: >//[self presentViewController:controller animated:YES completion:nil]; [](start = 4, length = 70)

Remove commented-out code.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    __weak UIView<ACRIContentHoldingView> *_superview;
+    BOOL isInline;
+}
+// dedicated initializer
+- (instancetype)initWithMediaEvent:(ACOMediaEvent *)mediaEvent
+                               url:(NSURL *)url
+                          rootView:(ACRView *)rootView
+                            config:(ACOHostConfig *)config
+                    containingview:(UIView *)containingview
+                         superview:(UIView<ACRIContentHoldingView> *)superview
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: I'm having trouble following this. The media target has a root view, a containing view, a superview, and optionally also a parent view controller provided by the media delegate. Does it need all these containers? The superview, for one, seems to be unused.

Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: Ben, thanks for the feedback; yes, superview is not used; its use was removed after i made changes to rely on ios api and tag to access the superview. root view provides context of adaptive cards, containing view is the superview that holds poster and media views. parent view controller is removed.

---
In reply to: [207006216](https://github.com/Microsoft/AdaptiveCards/pull/1736#discussion_r207006216) [](ancestors = 207006216)

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ViewController.m [hunk 14]
```diff
@@ -280,6 +281,12 @@ - (void) didFetchUserResponses:(ACOAdaptiveCard *)card action:(ACOBaseActionElem
 - (void)didFetchSecondaryView:(ACOAdaptiveCard *)card navigationController:(UINavigationController *)navigationController{
     [self presentViewController:navigationController animated:YES completion:nil];
 }
+
+- (UIViewController *)didFetchMediaViewController:(AVPlayerViewController *)controller card:(ACOAdaptiveCard *)card {
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ViewController.m]: >UIViewController * [](start = 3, length = 18)

This should be `void` now.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    }];
+}
+
+- (void)playMedia:(AVAssetTrack *)track asset:(AVURLAsset *)asset
+{
+    // video is ready to play, config AVPlayerViewController view dimension
+    CGSize size = track.naturalSize;
+    AVPlayerItem *item = [AVPlayerItem playerItemWithAsset:asset];
+    AVPlayer *player = [AVPlayer playerWithPlayerItem:item];
+    UIViewController *parentViewController = nil;
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm]: >UIViewController *parentViewController = nil; [](start = 4, length = 45)

This is unused now.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.h [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+- (instancetype)initWithMediaEvent:(ACOMediaEvent *)mediaEvent
+                          rootView:(ACRView *)rootView
+                            config:(ACOHostConfig *)config
+                    containingview:(UIView *)containingview;
+- (void)doSelectAction;
+
+- (void)playMedia:(AVAssetTrack *)track asset:(AVURLAsset *)asset;
+
+- (void)getAVTrack:(AVURLAsset *)asset;
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.h]: >getAVTrack [](start = 8, length = 10)

This method should also be renamed to indicate that it's playing the video and not just retrieving it.

**Pull request merged**